File: docs/js/simulate.js
/**
 * Monte Carlo pool simulator. Port of pipeline/simulate.py; keep in sync.
 *
 * The field's trajectory is independent of my pick, so it is drawn once and
 * every candidate path is scored against the same outcomes.
 */

/** mulberry32: small, fast, seedable. */
export function rng(seed) {
  let a = seed >>> 0;
  return () => {
    a = (a + 0x6D2B79F5) >>> 0;
    let t = a;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function gaussian(rand) {
  let u = 0;
  while (u === 0) u = rand();
  return Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * rand());
}

export function binomial(n, p, rand) {
  if (n <= 0 || p <= 0) return 0;
  if (p >= 1) return n;
  if (n <= 64) {
    let k = 0;
    for (let i = 0; i < n; i++) if (rand() < p) k++;
    return k;
  }
  const mean = n * p;
  const sd = Math.sqrt(n * p * (1 - p));
  const k = Math.round(mean + sd * gaussian(rand));
  return Math.min(n, Math.max(0, k));
}

export function prepare(byWeek, weeks) {
  return weeks.map((w) => {
    const wg = byWeek.get(w) || [];
    const key = (g) => [g.team, g.opp].sort().join('|');
    const gameIdx = new Map();
    const p0 = [];
    for (const g of wg) {
      const k = key(g);
      if (!gameIdx.has(k)) {
        gameIdx.set(k, p0.length);
        const first = k.split('|')[0];
        p0.push(g.team === first ? g.p : 1 - g.p);
      }
    }
    let tot = 0;
    for (const g of wg) tot += g.pick || 0;
    const teams = wg.map((g) => ({
      team: g.team,
      gi: gameIdx.get(key(g)),
      side: g.team === key(g).split('|')[0] ? 0 : 1,
      share: tot > 0 ? (g.pick || 0) / tot : 0,
    }));
    return { week: w, p0: Float64Array.from(p0), teams, byTeam: new Map(teams.map((t) => [t.team, t])) };
  });
}

/**
 * @param {object} opts  lives: strikes allowed before elimination (1 = single elimination);
 *                       fieldStrikes: entries in the field already sitting on their last life.
 */
export function simulateField(prep, poolSize, nSims, rand, { lives = 1, fieldStrikes = 0 } = {}) {
  const W = prep.length;
  const won = [];
  const alive = new Array(W + 1);
  const field = Math.max(0, poolSize - 1);
  // bucket[k][s] = entries with k strikes in sim s
  let bucket = [];
  for (let k = 0; k < lives; k++) bucket.push(new Int32Array(nSims));
  const onLast = lives > 1 ? Math.min(fieldStrikes, field) : 0;
  bucket[0].fill(field - onLast);
  if (lives > 1) bucket[lives - 1].fill(onLast);
  alive[0] = new Int32Array(nSims).fill(field);
  const frac = new Float64Array(nSims);
  for (let i = 0; i < W; i++) {
    const pw = prep[i];
    const G = pw.p0.length;
    const out = new Uint8Array(nSims * G);
    for (let s = 0; s < nSims; s++) {
      for (let g = 0; g < G; g++) out[s * G + g] = rand() < pw.p0[g] ? 1 : 0;
    }
    won.push(out);
    frac.fill(0);
    for (const t of pw.teams) {
      if (!t.share) continue;
      for (let s = 0; s < nSims; s++) {
        const w = out[s * G + t.gi] ^ t.side;
        if (w) frac[s] += t.share;
      }
    }
    const next = bucket.map(() => new Int32Array(nSims));
    const total = new Int32Array(nSims);
    for (let s = 0; s < nSims; s++) {
      const f = Math.min(1, frac[s]);
      for (let k = 0; k < lives; k++) {
        const n = bucket[k][s];
        if (!n) continue;
        const surv = binomial(n, f, rand);
        next[k][s] += surv;
        if (k + 1 < lives) next[k + 1][s] += n - surv;   // losers spend a life
      }
      for (let k = 0; k < lives; k++) total[s] += next[k][s];
    }
    bucket = next;
    alive[i + 1] = total;
  }
  return { won, alive, nSims, lives };
}

export function scorePath(prep, field, path, { lives = 1, myStrikes = 0 } = {}) {
  const { won, alive, nSims } = field;
  const W = prep.length;
  const weekIdx = new Map(prep.map((pw, i) => [pw.week, i]));
  const meAlive = new Uint8Array(nSims).fill(1);
  const strikes = new Uint8Array(nSims).fill(Math.min(myStrikes, lives - 1));
  const settled = new Uint8Array(nSims);
  const payout = new Float64Array(nSims);
  const curve = [];
  for (const { week, team } of path) {
    const i = weekIdx.get(week);
    const pw = prep[i];
    const t = pw.byTeam.get(team);
    const G = pw.p0.length;
    const before = alive[i];
    const after = alive[i + 1];
    let aliveCount = 0;
    for (let s = 0; s < nSims; s++) {
      const myWin = won[i][s * G + t.gi] ^ t.side;
      if (meAlive[s] && !myWin) {
        strikes[s]++;
        if (strikes[s] >= lives) {
          meAlive[s] = 0;
          if (!settled[s]) {
            if (after[s] === 0) payout[s] = 1 / (before[s] + 1);
            settled[s] = 1;
          }
        }
      }
      if (!settled[s] && meAlive[s] && after[s] === 0) { payout[s] = 1; settled[s] = 1; }
      aliveCount += meAlive[s];
    }
    curve.push(aliveCount / nSims);
  }
  let outright = 0;
  let sum = 0;
  let clean = 0;
  for (let s = 0; s < nSims; s++) {
    if (!settled[s] && meAlive[s]) payout[s] = 1 / (alive[W][s] + 1);
    sum += payout[s];
    if (payout[s] === 1) outright++;
    if (meAlive[s] && strikes[s] === 0) clean++;
  }
  return {
    equity: sum / nSims,
    pWinOutright: outright / nSims,
    pSurvive: curve.length ? curve[curve.length - 1] : 1,
    pClean: clean / nSims,
    curve,
  };
}

export function expectedSurvivors(field) {
  return field.alive.slice(1).map((a) => {
    let s = 0;
    for (let i = 0; i < a.length; i++) s += a[i];
    return s / a.length + 1;
  });
}

File: pipeline/simulate.py
"""Reference Monte Carlo pool simulator (Python, numpy).

Model
-----
* Each remaining week, every game is a Bernoulli draw with the favorite's
  probability.  One draw per game, so both sides are consistent.
* The field (pool size − 1 entries) picks teams in proportion to that
  week's pick shares.  Survivors are drawn ``Binomial(alive, Σ share·won)``,
  which captures small-pool variance.
* The field's trajectory does not depend on *my* pick, so it is simulated
  once and every candidate path is scored against the same outcomes
  (common random numbers ⇒ low-variance comparisons).
* Multi-life pools: each entry (mine and the field's) may absorb
  ``lives − 1`` losses; losers move to the next strike bucket.
* Payout: I win 1 if I'm the last entry standing; if the season ends with
  S survivors (me included) I take 1/S; if every remaining entry loses in
  the same week, that week's entrants split the pot.

The browser port in ``docs/js/simulate.js`` follows this exactly.
"""
from __future__ import annotations

import numpy as np


def prepare(games: list[dict], weeks: list[int]) -> dict:
    """Index games into per-week arrays: unique games, team→(game idx, side)."""
    per_week = []
    for w in weeks:
        wg = [g for g in games if g["week"] == w]
        seen = {}
        entries = []
        for g in wg:
            key = tuple(sorted((g["team"], g["opp"])))
            if key not in seen:
                seen[key] = len(entries)
                # probability that key[0] wins
                p0 = g["p"] if g["team"] == key[0] else 1 - g["p"]
                entries.append(p0)
        team_side = {}
        for g in wg:
            key = tuple(sorted((g["team"], g["opp"])))
            team_side[g["team"]] = (seen[key], 0 if g["team"] == key[0] else 1)
        shares = {g["team"]: (g.get("pick") or 0.0) for g in wg}
        tot = sum(shares.values()) or 1.0
        shares = {t: s / tot for t, s in shares.items()}
        per_week.append({"p0": np.array(entries), "team_side": team_side, "shares": shares})
    return {"weeks": weeks, "per_week": per_week}


def simulate_field(prep: dict, pool_size: int, n_sims: int, rng: np.random.Generator,
                   lives: int = 1, field_strikes: int = 0) -> dict:
    """Draw outcomes and the field's survivor counts for every sim/week.

    ``lives`` is the number of losses that eliminate an entry (1 = single
    elimination).  ``field_strikes`` is how many entries already sit on their
    last life when the simulation starts.
    """
    W = len(prep["weeks"])
    won = []
    field = max(0, pool_size - 1)
    on_last = min(field_strikes, field) if lives > 1 else 0
    bucket = np.zeros((lives, n_sims), dtype=np.int64)      # entries by strikes used
    bucket[0] = field - on_last
    if lives > 1:
        bucket[lives - 1] = on_last
    alive = np.zeros((W + 1, n_sims), dtype=np.int64)
    alive[0] = field
    for i, pw in enumerate(prep["per_week"]):
        outcome = rng.random((n_sims, len(pw["p0"]))) < pw["p0"]
        won.append(outcome)
        frac = np.zeros(n_sims)
        for team, (gi, side) in pw["team_side"].items():
            team_won = outcome[:, gi] if side == 0 else ~outcome[:, gi]
            frac += pw["shares"].get(team, 0.0) * team_won
        frac = np.clip(frac, 0, 1)
        nxt = np.zeros_like(bucket)
        for k in range(lives):
            surv = rng.binomial(bucket[k], frac)
            nxt[k] += surv
            if k + 1 < lives:
                nxt[k + 1] += bucket[k] - surv
        bucket = nxt
        alive[i + 1] = bucket.sum(axis=0)
    return {"won": won, "alive": alive, "lives": lives}


def score_path(prep: dict, field: dict, path: list[tuple[int, str]],
               lives: int = 1, my_strikes: int = 0) -> dict:
    W = len(prep["weeks"])
    n = field["alive"].shape[1]
    week_idx = {w: i for i, w in enumerate(prep["weeks"])}
    me_alive = np.ones(n, dtype=bool)
    strikes = np.full(n, min(my_strikes, lives - 1), dtype=np.int64)
    payout = np.zeros(n)
    settled = np.zeros(n, dtype=bool)
    survive_curve = []
    for w, team in path:
        i = week_idx[w]
        pw = prep["per_week"][i]
        gi, side = pw["team_side"][team]
        my_win = field["won"][i][:, gi] if side == 0 else ~field["won"][i][:, gi]
        before = field["alive"][i]
        after = field["alive"][i + 1]
        lost = me_alive & ~my_win
        strikes[lost] += 1
        out_now = lost & (strikes >= lives)
        everyone_out = out_now & ~settled & (after == 0)
        payout[everyone_out] = 1.0 / (before[everyone_out] + 1)
        settled |= out_now
        me_alive &= ~out_now
        last_standing = ~settled & me_alive & (after == 0)
        payout[last_standing] = 1.0
        settled |= last_standing
        survive_curve.append(me_alive.mean())
    end = ~settled & me_alive
    payout[end] = 1.0 / (field["alive"][W][end] + 1)
    return {
        "equity": float(payout.mean()),
        "p_win_outright": float((payout == 1.0).mean()),
        "p_survive": float(me_alive.mean()),
        "p_clean": float((me_alive & (strikes == 0)).mean()),
        "survive_curve": survive_curve,
        "expected_survivors": (field["alive"][1:].mean(axis=1) + 1).tolist(),
    }
